skfda/exploratory/visualization/representation.py

```diff
@@ -1,32 +1,70 @@
+"""Representation Module.
+
+This module contains the functionality related
+with plotting and scattering our different datasets.
+It allows multiple modes and colors, which could
+be set manually or automatically depending on values
+like depth measures.
+"""
+
+from typing import (
+    Any,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Tuple,
+    TypeVar,
+    Union,
+)
 
 import matplotlib.cm
 import matplotlib.patches
 import numpy as np
+from matplotlib.axes import Axes
+from matplotlib.figure import Figure
 
+from ... import FDataGrid
 from ..._utils import _to_domain_range, constants
+from ...representation._functional_data import FData
+from ...representation._typing import DomainRangeLike
 from ._utils import (
     _get_figure_and_axes,
     _set_figure_layout_for_fdata,
     _set_labels,
 )
 
+K = TypeVar('K')
+T = TypeVar('T', FDataGrid, np.ndarray)
 
-def _get_label_colors(n_labels, group_colors=None):
-    """Get the colors of each label"""
 
+def _get_label_colors(
+    n_labels: int,
+    group_colors: Union[Sequence[Any], Mapping[K, Any], None],
+) -> np.ndarray:
+    """Get the colors of each label."""
     if group_colors is not None:
         if len(group_colors) != n_labels:
-            raise ValueError("There must be a color in group_colors "
-                             "for each of the labels that appear in "
-                             "group.")
+            raise ValueError(
+                "There must be a color in group_colors "
+                "for each of the labels that appear in "
+                "group.",
+            )
     else:
         colormap = matplotlib.cm.get_cmap()
         group_colors = colormap(np.arange(n_labels) / (n_labels - 1))
 
     return group_colors
 
 
-def _get_color_info(fdata, group, group_names, group_colors, legend, kwargs):
+def _get_color_info(
+    fdata: T,
+    group: Union[Sequence[Any], None],
```

Comment:
This is a though function to type. Ideally the `@overload` decorator could be used. In that case you would have three overloads:
- One in which group is `None` and then `group_names` and `group_colors` must be also `None`.
- One in which group is `Sequence[int]` and `group_names` is `Union[Sequence[str], Mapping[K, str], None]` and `group_colors` is `Union[Sequence[Any], Mapping[K, Any], None]` (or a better type for the colors if you can find it).
- One in which group is `Sequence[K]` and `group_names` is `Optional[Mapping[K, str]]` and `group_colors` is `Optional[Mapping[K, Any]]`.

The problem with that approach is that currently the implementation is not checked against every overload (see https://github.com/python/mypy/issues/8867), so you woud probably have mypy errors.

Another possibility, that it is arguably better, as it requires less from the arguments is to define and use a [protocol](https://mypy.readthedocs.io/en/stable/protocols.html). You could define a generic protocol `Indexable[K, V]` and then use `Optional[Indexable[K, str]]` for `group_labels`, `Optional[Indexable[K, Any]]` for `group_colors` and `Optional[Sequence[K]]` for `group`.

Reply:
i think i corrected it, i finally went for the overload